common/flatpak-transaction-log.c

```diff
@@ -0,0 +1,293 @@
+{
+  g_autoptr(GFile) parent = g_file_get_parent (file);
+  g_autoptr(GFileOutputStream) stream = NULL;
+  g_autofree gchar *serialized = NULL;
+  gsize serialized_len = 0;
+  gsize written = 0;
+  g_autoptr(JsonGenerator) generator = NULL;
+  g_autoptr(GError) my_error = NULL;
+
+  if (!g_file_make_directory_with_parents (parent, cancellable, &my_error))
+    {
+      if (!g_error_matches (my_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        {
+          g_propagate_error (error, g_steal_pointer (&my_error));
+          return FALSE;
+        }
+
+      g_clear_error (&my_error);
+    }
+
+  generator = json_generator_new ();
+  json_generator_set_root (generator, node);
+  json_generator_set_pretty (generator, FALSE);
+
+  /* Resize the buffer to accomodate one extra newline and write that
+   * in place at serialized_len */
+  serialized = json_generator_to_data (generator, &serialized_len);
+  serialized = g_realloc (serialized, (serialized_len + 1) * sizeof (gchar));
+  serialized[serialized_len++] = '\n';
+  serialized[serialized_len] = '\0';
+
+  stream = g_file_append_to (file,
+                             G_FILE_CREATE_NONE,
+                             cancellable,
+                             error);
+
+  return g_output_stream_write_all (G_OUTPUT_STREAM (stream),
+                                    (gconstpointer) serialized,
+                                    serialized_len,
+                                    &written,
+                                    cancellable,
+                                    error);
+}
+
+static JsonNode *
+build_base_event (const gchar *event)
+{
+  g_autoptr(JsonBuilder) builder = json_builder_new ();
+
+  json_builder_begin_object (builder);
+
+  json_builder_set_member_name (builder, "event");
+  json_builder_add_string_value (builder, event);
+
+  json_builder_end_object (builder);
+
+  return json_builder_get_root (builder);
+}
+
+static JsonNode *
+deploy_event (const gchar *ref,
+              const gchar *origin,
+              const gchar *commit)
+{
+  g_autoptr(JsonNode) base = build_base_event ("deploy");
+  JsonObject *object = json_node_get_object (base);
+
+  json_object_set_string_member (object, "ref", ref);
+  json_object_set_string_member (object, "origin", origin);
+  json_object_set_string_member (object, "commit", commit);
+
+  return g_steal_pointer (&base);
+}
+
+/**
+ * flatpak_transaction_log_commit_deploy_event:
+ * @log: a #FlatpakTransactionLog
+ * @ref: a decomposable Flatpak ref
+ * @origin: the origin of the flatpak to be installed, whether it is from
+ *          a bundle or from a remote.
+ * @cancellable: a #GCancellable
+ * @error: a #GError
+ *
+ * Write a "deploy-install" event to the transaction log. The "ref" and
+ * "origin" arguments will be included in the event description as properties.
+ *
+ * Since: 0.10.0
+ * Returns: %TRUE if writing the event succeeded.
+ */
+gboolean
+flatpak_transaction_log_commit_deploy_event (FlatpakTransactionLog *log,
```

Comment:
Nitpick: By convention, method implementations for GObject classes always call their first parameter `self`.